File: backend/app/core/checkout_elements.py
"""Single source of truth for which optional elements the Billing checkout screen can show, and
what a tenant is allowed to turn on/off. Settings > Billing Settings > Checkout Elements reads
this catalog to render its toggles; POSPage reads the resolved per-tenant config to decide what to
render. Adding a future element (Coupon, Loyalty, Gift Card, Service Charge, Table, Waiter, ...) is
just one more entry here — no schema migration, no new endpoint, no scattered conditionals to hunt
down elsewhere.

Total and Checkout are deliberately NOT in this registry — they're mandatory and always render;
giving them a toggle would let a tenant configure away the ability to complete a sale.
"""
from __future__ import annotations

from typing import Literal, TypedDict

CheckoutElementGroup = Literal["customer_info", "pricing", "payment_type", "payment_method", "cashier_tools"]

GROUP_LABELS: dict[CheckoutElementGroup, str] = {
    "customer_info": "Customer Info",
    "pricing": "Pricing",
    "payment_type": "Payment Type",
    "payment_method": "Payment Method",
    "cashier_tools": "Cashier Tools",
}


class CheckoutElement(TypedDict):
    key: str
    label: str
    group: CheckoutElementGroup
    default: bool
    # Module key (see app/core/feature_catalog.py) this element additionally depends on, if any.
    # The element only ever shows when BOTH this module is enabled for the tenant AND the
    # tenant's own checkout-element setting for it is on.
    depends_on_module: str | None


def _e(
    key: str, label: str, group: CheckoutElementGroup, *, default: bool = True, depends_on_module: str | None = None
) -> CheckoutElement:
    return {"key": key, "label": label, "group": group, "default": default, "depends_on_module": depends_on_module}


# Safe defaults: every element defaults to True, matching the app's existing hard-coded-on
# behavior before this feature existed — a tenant who never visits the new settings page keeps
# exactly the checkout they have today.
CHECKOUT_ELEMENT_REGISTRY: list[CheckoutElement] = [
    _e("customer", "Customer", "customer_info"),
    _e("phone", "Phone", "customer_info"),
    _e("discount", "Discount", "pricing"),
    _e("tax", "Tax", "pricing"),
    _e("paid_in_full", "Paid in Full", "payment_type"),
    _e("partially_paid", "Partially Paid", "payment_type", depends_on_module="payments_credit"),
    _e("credit", "Credit (Pay Later)", "payment_type", depends_on_module="payments_credit"),
    _e("cash", "Cash", "payment_method"),
    _e("card", "Card", "payment_method"),
    _e("upi", "UPI", "payment_method"),
    _e("amount_tendered", "Amount Tendered", "cashier_tools"),
    _e("change_due", "Change Due", "cashier_tools"),
    _e("hold_bill", "Hold Bill", "cashier_tools"),
]

CHECKOUT_ELEMENT_BY_KEY: dict[str, CheckoutElement] = {e["key"]: e for e in CHECKOUT_ELEMENT_REGISTRY}
CHECKOUT_ELEMENT_KEYS: set[str] = set(CHECKOUT_ELEMENT_BY_KEY)

DEFAULT_CHECKOUT_CONFIG: dict[str, bool] = {e["key"]: e["default"] for e in CHECKOUT_ELEMENT_REGISTRY}


def resolve_checkout_config(stored: dict[str, bool] | None) -> dict[str, bool]:
    """Merges a tenant's stored overrides (possibly missing keys — an old tenant predating a
    newly added element, or one that's never touched this page at all) with registry defaults.
    Unknown keys in `stored` (e.g. a since-removed element) are dropped rather than surfaced."""
    stored = stored or {}
    return {key: stored.get(key, default) for key, default in DEFAULT_CHECKOUT_CONFIG.items()}

File: billing-app/src/features/pos/lib/checkoutElements.ts
import { useQuery } from '@tanstack/react-query';
import * as settingsApi from '@/features/settings/api';

/** The checkout elements POSPage/CheckoutPanel have bespoke rendering for today. Mirrors (but
 * does not replace) the backend registry in app/core/checkout_elements.py — that catalog is the
 * source of truth for defaults/labels/dependencies and drives the Settings toggle list generically;
 * this union just gives the checkout page's specific per-element conditionals compile-time safety.
 * Total and Checkout are deliberately absent — mandatory, never configurable. */
export type CheckoutElementKey =
  | 'customer'
  | 'phone'
  | 'discount'
  | 'tax'
  | 'paid_in_full'
  | 'partially_paid'
  | 'credit'
  | 'cash'
  | 'card'
  | 'upi'
  | 'payment_reference'
  | 'amount_tendered'
  | 'change_due'
  | 'hold_bill'
  | 'table'
  | 'print_order_bill'
  | 'print_kot';

/** Same safe-default philosophy as the backend (app/core/checkout_elements.py
 * DEFAULT_CHECKOUT_CONFIG) and useFeatureFlag: absent/still-loading reads as enabled, so a tenant
 * who's never configured this — or a slow network on first paint — sees today's full checkout,
 * never one that's silently missing pieces. */
export const DEFAULT_CHECKOUT_CONFIG: Record<CheckoutElementKey, boolean> = {
  customer: true,
  phone: true,
  discount: true,
  tax: true,
  paid_in_full: true,
  partially_paid: true,
  credit: true,
  cash: true,
  card: true,
  upi: true,
  payment_reference: true,
  amount_tendered: true,
  change_due: true,
  hold_bill: true,
  table: true,
  print_order_bill: true,
  print_kot: true,
};

/** Mirrors app/core/checkout_elements.py's group assignment — used to compute the layout reflow
 * (see checkoutLayout.ts), not for rendering per-element UI (that's bespoke per element already). */
export const ELEMENT_GROUPS: Record<CheckoutElementKey, string> = {
  customer: 'customer_info',
  phone: 'customer_info',
  discount: 'pricing',
  tax: 'pricing',
  paid_in_full: 'payment_type',
  partially_paid: 'payment_type',
  credit: 'payment_type',
  cash: 'payment_method',
  card: 'payment_method',
  upi: 'payment_method',
  payment_reference: 'payment_method',
  amount_tendered: 'cashier_tools',
  change_due: 'cashier_tools',
  hold_bill: 'cashier_tools',
  table: 'cashier_tools',
  print_order_bill: 'cashier_tools',
  print_kot: 'cashier_tools',
};

export function useCheckoutConfigQuery() {
  return useQuery({
    queryKey: ['checkout-config'],
    queryFn: settingsApi.getCheckoutConfig,
    staleTime: 60_000,
  });
}

/** Resolved per-element visibility for the current tenant — absent keys (not yet fetched, or a
 * newly added element an older cached response predates) fall back to enabled. */
export function useCheckoutConfig(): Record<CheckoutElementKey, boolean> {
  const { data } = useCheckoutConfigQuery();
  const config = data?.config;
  const result = { ...DEFAULT_CHECKOUT_CONFIG };
  if (config) {
    for (const key of Object.keys(result) as CheckoutElementKey[]) {
      if (config[key] !== undefined) result[key] = config[key];
    }
  }
  return result;
}
